File: backend/app/api/routes/interviews.py
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, File, UploadFile, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.audio.assemblyai_service import AssemblyAIService, validate_audio_file
from app.db.session import get_db_session
from app.schemas.interview import InterviewCreate, InterviewDetailRead, InterviewRead
from app.schemas.template import (
    AnalyzeCVResponse,
    AudioAnswerResponse,
    AnswerRequest,
    AnswerTurnResponse,
    InterviewCreateV2,
    InterviewFinalReport,
    InterviewScoreResponse,
    StartInterviewResponse,
)
from app.services import interview_service, template_interview_service

router = APIRouter(prefix="/interviews", tags=["interviews"])


@router.post(
    "",
    response_model=InterviewRead,
    status_code=status.HTTP_201_CREATED,
)
async def create_interview(
    payload: InterviewCreate | InterviewCreateV2,
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    if isinstance(payload, InterviewCreateV2):
        return await template_interview_service.create_interview_from_template(session, payload)
    return await interview_service.create_interview(session, payload)


@router.get("", response_model=list[InterviewRead])
async def list_interviews(
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    return await interview_service.list_interviews(session)


@router.get("/{interview_id}", response_model=InterviewDetailRead)
async def get_interview(
    interview_id: uuid.UUID,
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    return await interview_service.get_interview_detail(session, interview_id)


@router.post("/{interview_id}/upload-cv")
async def upload_cv_for_interview(
    interview_id: uuid.UUID,
    file: Annotated[UploadFile, File(...)],
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    return await template_interview_service.upload_cv(session, interview_id, file)


@router.post("/{interview_id}/analyze-cv", response_model=AnalyzeCVResponse)
async def analyze_cv_for_interview(
    interview_id: uuid.UUID,
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    return await template_interview_service.analyze_cv_against_template(session, interview_id)


@router.post("/{interview_id}/send-candidate-invite", response_model=InterviewRead)
async def send_candidate_invite(
    interview_id: uuid.UUID,
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    return await template_interview_service.send_candidate_invite(session, interview_id)


@router.post("/{interview_id}/start", response_model=StartInterviewResponse)
async def start_template_interview(
    interview_id: uuid.UUID,
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    return await template_interview_service.start_template_interview(session, interview_id)


@router.post("/{interview_id}/answer", response_model=AnswerTurnResponse)
async def answer_template_interview(
    interview_id: uuid.UUID,
    payload: AnswerRequest,
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    result = await template_interview_service.submit_text_answer(
        session,
        interview_id,
        payload.answer,
    )
    return AnswerTurnResponse(interview_id=interview_id, **result)


@router.post("/{interview_id}/audio-answer", response_model=AudioAnswerResponse)
async def answer_template_interview_with_audio(
    interview_id: uuid.UUID,
    file: Annotated[UploadFile, File(...)],
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    validate_audio_file(file)
    transcript = await AssemblyAIService().transcribe_upload(file)
    result = await template_interview_service.submit_text_answer(
        session,
        interview_id,
        transcript,
    )
    score = await template_interview_service.get_score(session, interview_id)
    return AudioAnswerResponse(
        candidate_transcript=result["candidate_transcript"],
        evaluation=result["evaluation"],
        current_score={
            "initial_cv_score": score["initial_cv_score"],
            "question_score": score["question_score"],
            "bonus_score": score["bonus_score"],
            "final_score": score["final_score"],
            "max_score": score["max_score"],
            "percentage": score["percentage"],
        },
        next_question={
            "id": result.get("next_question_id"),
            "question_text": result.get("next_question"),
        },
        interview_status=result["status"],
    )


@router.post("/{interview_id}/finalize", response_model=InterviewFinalReport)
async def finalize_template_interview(
    interview_id: uuid.UUID,
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    return await template_interview_service.finalize_interview(session, interview_id)


@router.get("/{interview_id}/score", response_model=InterviewScoreResponse)
async def get_template_interview_score(
    interview_id: uuid.UUID,
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    return await template_interview_service.get_score(session, interview_id)


@router.get("/{interview_id}/report", response_model=InterviewFinalReport)
async def get_template_interview_report(
    interview_id: uuid.UUID,
    session: Annotated[AsyncSession, Depends(get_db_session)],
):
    return await template_interview_service.get_final_report(session, interview_id)

File: frontend/lib/services/interviews.ts
import { apiRequest } from "@/lib/api";
import type {
  AudioAnswerResponse,
  AnalyzeCVResponse,
  AnswerTurnResponse,
  CandidateReport,
  Interview,
  InterviewCreate,
  InterviewCreateFromTemplate,
  InterviewDetail,
  InterviewScore,
  StartInterviewResponse,
} from "@/types/api";

export async function createInterview(payload: InterviewCreate): Promise<Interview> {
  return apiRequest<Interview>("/interviews", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function createInterviewFromTemplate(payload: InterviewCreateFromTemplate): Promise<Interview> {
  return apiRequest<Interview>("/interviews", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function listInterviews(): Promise<Interview[]> {
  return apiRequest<Interview[]>("/interviews");
}

export async function getInterview(interviewId: string): Promise<InterviewDetail> {
  return apiRequest<InterviewDetail>(`/interviews/${interviewId}`);
}

export async function uploadInterviewCv(interviewId: string, file: File) {
  const form = new FormData();
  form.append("file", file);
  return apiRequest(`/interviews/${interviewId}/upload-cv`, {
    method: "POST",
    body: form,
  });
}

export async function analyzeInterviewCv(interviewId: string) {
  return apiRequest<AnalyzeCVResponse>(`/interviews/${interviewId}/analyze-cv`, {
    method: "POST",
  });
}

export async function sendCandidateInvite(interviewId: string): Promise<Interview> {
  return apiRequest<Interview>(`/interviews/${interviewId}/send-candidate-invite`, {
    method: "POST",
  });
}

export async function startTemplateInterview(interviewId: string): Promise<StartInterviewResponse> {
  return apiRequest<StartInterviewResponse>(`/interviews/${interviewId}/start`, {
    method: "POST",
  });
}

export async function answerTemplateInterview(interviewId: string, answer: string): Promise<AnswerTurnResponse> {
  return apiRequest<AnswerTurnResponse>(`/interviews/${interviewId}/answer`, {
    method: "POST",
    body: JSON.stringify({ answer }),
  });
}

export async function answerTemplateInterviewAudio(interviewId: string, audioBlob: Blob): Promise<AudioAnswerResponse> {
  const extension = mimeToExtension(audioBlob.type);
  const form = new FormData();
  form.append("file", audioBlob, `answer.${extension}`);
  return apiRequest<AudioAnswerResponse>(`/interviews/${interviewId}/audio-answer`, {
    method: "POST",
    body: form,
  });
}

export async function finalizeTemplateInterview(interviewId: string): Promise<CandidateReport> {
  return apiRequest<CandidateReport>(`/interviews/${interviewId}/finalize`, {
    method: "POST",
  });
}

export async function getInterviewScore(interviewId: string): Promise<InterviewScore> {
  return apiRequest<InterviewScore>(`/interviews/${interviewId}/score`);
}

export async function getInterviewReport(interviewId: string): Promise<CandidateReport> {
  return apiRequest<CandidateReport>(`/interviews/${interviewId}/report`);
}

export async function updateInterviewAnswerScore(
  interviewId: string,
  answerId: string,
  finalQuestionScore: number,
): Promise<CandidateReport["answer_evaluations"][number]> {
  return apiRequest<CandidateReport["answer_evaluations"][number]>(
    `/interviews/${interviewId}/answers/${answerId}/score`,
    {
      method: "PATCH",
      body: JSON.stringify({ final_question_score: finalQuestionScore }),
    },
  );
}

function mimeToExtension(mimeType: string): string {
  if (mimeType.includes("wav")) return "wav";
  if (mimeType.includes("mpeg") || mimeType.includes("mp3")) return "mp3";
  if (mimeType.includes("ogg")) return "ogg";
  if (mimeType.includes("mp4")) return "mp4";
  return "webm";
}
